client/comms/wsconn.go

```diff
@@ -0,0 +1,349 @@
+	conn.reqMtx.Lock()
+	defer conn.reqMtx.Unlock()
+	req := conn.req[id]
+	if req == nil {
+		return nil, fmt.Errorf("no request found for id %d", id)
+	}
+
+	delete(conn.req, id)
+
+	return req, nil
+}
+
+// NextID returns the next request id.
+func (conn *WsConn) NextID() uint64 {
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	scheme := "ws"
+
+	if conn.tlsCfg != nil {
+		scheme = "wss"
+	}
+
+	url := url.URL{
+		Scheme: scheme,
+		Host:   conn.cfg.Host,
+		Path:   conn.cfg.Path,
+	}
+
+	dialer := &websocket.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 10 * time.Second,
+	}
+
+	if conn.tlsCfg != nil {
+		dialer.TLSClientConfig = conn.tlsCfg
+	}
+
+	var err error
+	conn.ws, _, err = dialer.Dial(url.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	conn.setConnected(true)
+	conn.ws.SetPingHandler(func(string) error {
+		now := time.Now()
+		err := conn.ws.SetReadDeadline(now.Add(conn.cfg.PingWait))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{},
+			now.Add(conn.cfg.WriteWait))
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-ticker.C:
+			if !conn.isConnected() {
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v", err)
+				}
+
+				// Run the reconnect sync after reestablishing a connection.
+				if conn.cfg.ReconnectSync != nil {
+					conn.cfg.ReconnectSync()
+				}
+			}
+
+		case <-ctx.Done():
+			now := time.Now()
+			closeMsg := websocket.FormatCloseMessage(websocket.CloseMessage, "")
+			conn.writeMtx.Lock()
+			err := conn.ws.WriteControl(websocket.CloseMessage, closeMsg,
+				now.Add(conn.cfg.WriteWait))
+			conn.writeMtx.Unlock()
+			if err != nil {
+				log.Errorf("unable to close websocket connection: %v", err)
+			}
+			ticker.Stop()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	for {
+		// Only allow reads if there is an established connection.
+		if !conn.isConnected() {
+			time.Sleep(time.Millisecond * 500)
+			continue
+		}
+
+		msg := new(msgjson.Message)
+		err := conn.ws.ReadJSON(msg)
+		if err != nil {
+			if opErr, ok := err.(*net.OpError); ok {
+				if opErr.Op == "read" {
+					conn.setConnected(false)
+					continue
+				}
+			}
+
+			if websocket.IsCloseError(err, websocket.CloseAbnormalClosure,
+				websocket.CloseGoingAway, websocket.CloseNormalClosure) {
+				return
+			}
```

Comment:
This is confusing to me because you have two normal close error codes (going away and normal) and one abnormal.  But all other errors that would successfully assert to a `*websocket.CloseError` pass through even though they would be recognized by gorilla.

I thing what would work is:
1. `IsUnexpectedCloseError(err, CloseAbnormalClosure, {other unexpected codes})` with a log of the error and return.
followed by
2. `IsUnexpectedCloseError(err)` with no log, then return.

Reply:
Sorry I see why that is confusing, this is what I intended:

```
if websocket.IsCloseError(err, websocket.CloseGoingAway, websocket.CloseNormalClosure) {
	// Return when a close message is received.
	conn.cfg.Cancel()
	return
}

...

log.Errorf("unable to read msg: %v", err)
conn.cfg.Cancel()
return
```

This handles normal close errors as expected and logs all other close errors as errors.